### PR TITLE
wip: minimal client

### PR DESCRIPTION
**Pull request by shawnsarwar**
We wanted to move from the legacy client to something slimmer that could be driven by the Kernel's exposed OpenAPI definitions. This PR accomplishes that for the most part. Unfortunately CoreAPI which makes up the meat of the client has some issues with our API. As such, only CREATE and READ operations are currently supported. As those are the operations in the client that are currently used in Integration Tests and the Producer, it was decided   that we could release a limited client.

As part of this work, we've also removed the integration test suite's dependency on Mocker and Saladbar. Saladbar has been removed entirely as it's not being used or actively maintained. It will likely serve as a prototype for future schema library concepts but as it exists now it is vestigial. Mocker remains in the codebase, but  as it's no longer required for integration testing is no longer automatically built / distribuited.

### CODE REVIEW

File: aether-utils/aether-client/aether/client_test/fixtures/mappings/mapping.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+		[
+			"levels.vaccines[*].count",
+			"CurrentStock.count"
+		],
+		[
+			"report_datetime",
+			"CurrentStock.date"
+		]
+	]
+}
```
Comment (obdulia-losantos) [inline aether-utils/aether-client/aether/client_test/fixtures/mappings/mapping.json]: I think that `tabs` love you :stuck_out_tongue:

Comment (shawnsarwar) [reply aether-utils/aether-client/aether/client_test/fixtures/mappings/mapping.json]: You know what, I think it's because sometimes I use jsonlint...

File: aether-utils/aether-client/conf/pip/requirements.txt [hunk 2]
```diff
@@ -10,13 +10,27 @@
...
 pycodestyle==2.3.1
 pyflakes==1.6.0
-requests==2.18.4
-urllib3==1.22
+pytest==3.8.0
+pytest-runner==4.2
+requests==2.19.1
+six==1.11.0
+uritemplate==3.0.0
+urllib3==1.23
```
Comment (obdulia-losantos) [inline aether-utils/aether-client/conf/pip/requirements.txt]: Being `aether-client` a distributed package you don't need to upgrade dependencies, just something like `aether-common` has.
https://github.com/eHealthAfrica/aether/blob/develop/aether-common-module/conf/pip/requirements.txt

I'm just trying to reduce complexity :grimacing:

File: aether-utils/aether-client/entrypoint.sh [hunk 3]
```diff
@@ -41,21 +41,17 @@ show_help() {
...
-    coverage run    --rcfile="$RCFILE" manage.py test "${@:1}"
-    coverage report --rcfile="$RCFILE"
-    coverage erase
+test(){
+        python3 setup.py -q test "${@:1}"
 
-    cat /code/conf/extras/good_job.txt
+        cat /code/conf/extras/good_job.txt
+        rm -R ./*.egg*
+        rm -R .pytest_cache
```
Comment (obdulia-losantos) [inline aether-utils/aether-client/entrypoint.sh]: I think that 4 spaces were enough...

File: aether-utils/docker-compose.yml [hunk 4]
```diff
@@ -7,27 +7,20 @@ version: "2.1"
...
-    command: build
-
-
-  saladbar:
-    build: ./aether-saladbar
-    environment:
-      TESTING: "true"
-    volumes:
-      - ./aether-saladbar:/code
+      - ./aether-utils/aether-mock-data:/code
```
Comment (obdulia-losantos) [inline aether-utils/docker-compose.yml]: Not sure if with `./aether-utils` is going to work...

File: test-aether-integration-module/conf/pip/primary-requirements.txt [hunk 5]
```diff
@@ -11,8 +11,7 @@
 
 aet.consumer
 aether.client
-aether.mocker
-aether.saladbar
+coreapi
```
Comment (obdulia-losantos) [inline test-aether-integration-module/conf/pip/primary-requirements.txt]: This dependency must be inherited from `aether-client`, it's ok having it here but you can also include the comment ;-)

File: test-aether-integration-module/conf/pip/requirements.txt [hunk 6]
```diff
@@ -9,9 +9,7 @@
 #                                                                              		#
 #########################################################################################
 aet.consumer==0.9.1
-aether.client==0.0.0
-aether.mocker==0.0.0
-aether.saladbar==0.0.0
+aether.client==0.9.0
```
Comment (obdulia-losantos) [inline test-aether-integration-module/conf/pip/requirements.txt]: wondering how is this going to affect new releases...

**Review by obdulia-losantos — COMMENTED** [review]
not tested yet, just first :eyes:

File: aether-utils/aether-client/aether/client_test/fixtures/mappings/mapping.json (hunk 1 above)
Comment (obdulia-losantos) [reply aether-utils/aether-client/aether/client_test/fixtures/mappings/mapping.json]: You can indicate the `--indent` option :see_no_evil: 

https://github.com/zaach/jsonlint/

File: test-aether-integration-module/conf/pip/requirements.txt (hunk 6 above)
Comment (shawnsarwar) [reply test-aether-integration-module/conf/pip/requirements.txt]: I swapped it for >= in requirement.txt, but there's not way that I can find to get pip_freeze to generate anything but == in its output.

Comment (obdulia-losantos) [reply test-aether-integration-module/conf/pip/requirements.txt]: You can remove the dependency here, and introduce it in the `requirements_header.txt` file :thinking:

File: aether-utils/aether-client/conf/pip/requirements.txt (hunk 2 above)
Comment (shawnsarwar) [reply aether-utils/aether-client/conf/pip/requirements.txt]: That makes a lot of sense. Sometimes you just follow the pattern without checking to see if it's required...

File: scripts/test_travis.sh [hunk 7]
```diff
@@ -28,6 +28,7 @@ case "$1" in
     ;;
 
     integration)
+        ./scripts/build_aether_containers.sh
```
Comment (obdulia-losantos) [inline scripts/test_travis.sh]: maybe `./scripts/build_aether_utils_and_distribute.sh` ? Because you want it out of `test_integration.sh` file :thinking:

File: scripts/test_integration.sh [hunk 8]
```diff
@@ -30,16 +32,11 @@ DC_TEST="docker-compose -f docker-compose-test.yml"
...
 echo "_____________________________________________ TESTING"
 
-
-./scripts/build_aether_utils_and_distribute.sh
 $DC_TEST down
 
 echo "_____________________________________________ Starting database"
 $DC_TEST up -d db-test
 
-build_container kernel
```
Comment (obdulia-losantos) [inline scripts/test_integration.sh:40]: I think this is still a thing...

File: scripts/build_aether_utils_and_distribute.sh [hunk 9]
```diff
@@ -21,36 +21,25 @@
 set -Eeuo pipefail
```
Comment (obdulia-losantos) [inline scripts/build_aether_utils_and_distribute.sh:21]: this file can be renamed to `build_aether_client_and_distribute.sh`

File: scripts/generate-docker-compose-credentials.sh [hunk 10]
```diff
@@ -109,5 +109,6 @@ UI_ADMIN_USERNAME=admin
 UI_ADMIN_PASSWORD=$(gen_random_string)
 UI_DJANGO_SECRET_KEY=$(gen_random_string)
 UI_DB_PASSWORD=$(gen_random_string)
+VERSION=$( cat "VERSION")
```
Comment (obdulia-losantos) [inline scripts/generate-docker-compose-credentials.sh:112]: Could you move this to the top? This is not "UI" related ;-)

File: scripts/build_aether_utils_and_distribute.sh [hunk 11]
```diff
@@ -21,36 +21,25 @@
...
 
 DC_UTILS="docker-compose -f ./aether-utils/docker-compose.yml"
+VERSION=$(cat "VERSION")
 
 $DC_UTILS down
 
-UTILS=( client mocker saladbar)
+UTILS=( client )
```
Comment (obdulia-losantos) [inline scripts/build_aether_utils_and_distribute.sh]: Where are you building `mocker`?

File: aether-utils/aether-client/setup.py [hunk 12]
```diff
@@ -19,15 +19,17 @@
 import os
 from setuptools import setup, find_packages
 
-# allow setup.py to be run from any path
-os.chdir(os.path.normpath(os.path.join(os.path.abspath(__file__), os.pardir)))
-
 setup(
```
Comment (obdulia-losantos) [inline aether-utils/aether-client/setup.py:22]: Wouldn't this work for you?

```python
with open(os.path.join('./code', 'VERSION')) as version_file:
    version = version_file.read().strip()

setup(
   ...
   version=version,
   ...
)
```

In the `docker-compose.yml` file:
```yaml
  client:
    build: ./aether-client
    environment:
      TESTING: "true"
    volumes:
      - ./aether-client:/code
      - ../../VERSION:/code/VERSION
    command: build
```

Comment (obdulia-losantos) [reply aether-utils/aether-client/setup.py:22]: Using the `.env` file forces everybody to re-execute the generate script every time there is a new release, and we all know that we are so much lazy even for rebuilding the containers...

File: docker-compose-test.yml [hunk 13]
```diff
@@ -54,6 +54,22 @@ services:
       - "9000:9000"
     command: start
 
+  # ---------------------------------
```
Comment (obdulia-losantos) [inline docker-compose-test.yml:57]: I'm wondering if this must go to `./aether-utils/docker-compose.yml` file. Something like this:

```yaml
  # ---------------------------------
  # Aether Kernel Client TEST dependencies
  # ---------------------------------
  db-test:
    extends:
      file: ../docker-compose-test.yml
      service: db-test

  kernel-test:
    extends:
      file: ../docker-compose-test.yml
      service: kernel-test

  # ---------------------------------
  # Aether Kernel Client
  # ---------------------------------
   client-test:
    build: ./aether-client
    env_file:
      - ../../.env
    environment:
      # use a different database when testing the integration with another container
      KERNEL_URL: http://kernel-test:9000
      KERNEL_ADMIN_USERNAME: ${KERNEL_ADMIN_USERNAME}
      KERNEL_ADMIN_PASSWORD: ${KERNEL_ADMIN_PASSWORD}
      TESTING: "true"
      VERSION: ${VERSION}
    volumes:
      - ./aether-client:/code
    command: test
    depends_on:
      db-test:
        condition: service_healthy
      kernel-test:
        condition: service_healthy
```

Comment (shawnsarwar) [reply docker-compose-test.yml:57]: Right now the docker-compose-test space uses it's own default network. We could create `aether_test` or something and that would allow us to spread the config around. FWIW, I don't mind having all of the test config in one place though.

File: aether-utils/aether-client/entrypoint.sh [hunk 14]
```diff
@@ -69,30 +64,15 @@ case "$1" in
...
     test_lint)
         test_flake8
     ;;
 
-    test_coverage)
-        test_coverage "${@:2}"
-    ;;
-
     build)
         # remove previous build if needed
```
Comment (obdulia-losantos) [inline aether-utils/aether-client/entrypoint.sh:77]: I know that you cannot test the module without kernel but before building it you can just check the linting rules, at least this is going to detect unassigned variables, wrong indentation... stupid checks that save time searching for further errors.

File: aether-utils/aether-client/setup.py (hunk 12 above)
Comment (shawnsarwar) [reply aether-utils/aether-client/setup.py:22]: OOh, mounting the VERSION file is a great idea.

File: docker-compose-test.yml (hunk 13 above)
Comment (obdulia-losantos) [reply docker-compose-test.yml:57]: But remember to include it in any of the tests :grimacing:

File: scripts/aether_functions.sh [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+function create_readonly_user {
+    docker-compose run kernel eval python /code/sql/create_readonly_user.py
+    docker-compose kill
+}
+
+# Run function found at first command line arg
+CALL=$1
+# If there are arguments they are also accessible
+ARGS=${@:2}
+$CALL $ARGS
```
Comment (obdulia-losantos) [inline scripts/aether_functions.sh:91]: I think that this script file & related ones deserve their own PR, don't you think the same?

**Comment by shawnsarwar** [comment]
With the issues with update/delete not working, and now some mysterious issues in Travis with 104 (connection reset by host), I'm going to give another openapi library a try:
https://github.com/pyopenapi/pyswagger

**Comment by shawnsarwar** [comment]
Also changing this to a WIP for the time being so @obdulia-losantos can spend her time on other things. (Thanks!)

**Comment by shawnsarwar** [comment]
Features merged as part of #378
